### PR TITLE
Feature: Add UI for SCA error banner and re-authentication flow

### PR DESCRIPTION
**Pull request by waterim**
### Details

Add UI for SCA error banner and re-authentication flow

### Fixed Issues

$ https://github.com/Expensify/App/issues/46064
PROPOSAL: N/A

### Tests

NOTE: You need to connect to someone local backend to test live mode of the stripe

1. Navigate to `/settings/subscription/add-payment-card`
2. Fill the data with card which requires authentication from here https://docs.stripe.com/testing#regulatory-cards
3. Select GBP currency
3. Observe that modal appears correctly 
4. Click Fail
4. Observe error UI and button for "Authenticate payment"
5. Click on it and should redirect to "add-payment-card"
6. Modal should appear

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as tests

### QA Steps

None

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [x] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>

https://github.com/user-attachments/assets/165e54cd-cce4-434d-bbfd-df4eb5e43c2b

</details>

### CODE REVIEW

**Comment by blimpich** [comment]
Looked at this locally. I think we should apply these changes so that (A) the authenticate button only shows if the error banner also shows, and (B) so that the error banner will show even if the user is on a free trial and even if there is no amount owed. Lemme know what you think.

```diff
diff --git a/src/libs/SubscriptionUtils.ts b/src/libs/SubscriptionUtils.ts
index 0cc4e05f7b..8516d8e082 100644
--- a/src/libs/SubscriptionUtils.ts
+++ b/src/libs/SubscriptionUtils.ts
@@ -175,7 +175,7 @@ function hasAmountOwed(): boolean {
  * @returns Whether there is a card authentication error.
  */
 function hasCardAuthenticatedError() {
-    return stripeCustomerId?.status === 'authentication_required' && amountOwed === 0;
+    return stripeCustomerId?.status === 'authentication_required' && getAmountOwed() === 0;
 }
 
 /**
@@ -460,5 +460,6 @@ export {
     getCardForSubscriptionBilling,
     hasSubscriptionGreenDotInfo,
     hasRetryBillingError,
+    hasCardAuthenticatedError,
     PAYMENT_STATUS,
 };
diff --git a/src/pages/settings/Subscription/CardSection/CardSection.tsx b/src/pages/settings/Subscription/CardSection/CardSection.tsx
index df3c6809e1..a501c61a8b 100644
--- a/src/pages/settings/Subscription/CardSection/CardSection.tsx
+++ b/src/pages/settings/Subscription/CardSection/CardSection.tsx
@@ -32,6 +32,7 @@ import CardSectionDataEmpty from './CardSectionDataEmpty';
 import RequestEarlyCancellationMenuItem from './RequestEarlyCancellationMenuItem';
 import type {BillingStatusResult} from './utils';
 import CardSectionUtils from './utils';
+import { hasCardAuthenticatedError } from '@libs/SubscriptionUtils';
 
 function CardSection() {
     const [isRequestRefundModalVisible, setIsRequestRefundModalVisible] = useState(false);
@@ -85,7 +86,9 @@ function CardSection() {
         BillingBanner = <TrialStartedBillingBanner />;
     } else if (SubscriptionUtils.hasUserFreeTrialEnded()) {
         BillingBanner = <TrialEndedBillingBanner />;
-    } else if (billingStatus) {
+    }
+
+    if (billingStatus) {
         BillingBanner = (
             <SubscriptionBillingBanner
                 title={billingStatus.title}
@@ -144,7 +147,7 @@ function CardSection() {
                         large
                     />
                 )}
-                {billingStatus?.isAuthenticationRequired !== undefined && (
+                {hasCardAuthenticatedError() && (
                     <Button
                         text={translate('subscription.cardSection.authenticatePayment')}
                         isDisabled={isOffline || !billingStatus?.isAuthenticationRequired}
```

**Comment by blimpich** [comment]
> @blimpich Looks a bit weird this part: "getAmountOwed() === 0", do we need to check for owed amount for authentication error? Shouldn't it be only like this return stripeCustomerId?.status === 'authentication_required' ?

This confuses me too, and to be completely honest I don't understand why that is the logic, but that is the logic that we have in the Old Dot version of the application too, so we should try to mimic it. I looked into why that code was written that way in the old version of the app, but there wasn't really an explanation in the PR that added it, so 🤷. Lets keep it with the owed amount conditional.

My guess is that if there is an owed amount we have prioritized showing a different error banner instead of the authentication error banner.

**Comment by waterim** [comment]
@blimpich Can you please verify it works or not

**Comment by blimpich** [comment]
@waterim I will try to get to it today. I was on vacation yesterday so couldn't get to it.

**Comment by blimpich** [comment]
Error banner looks good but when I click "Authenticate" it just opens up a empty billing card form. It looks like the problem is that we clear the authenticationLink onyx key [here](https://github.com/Expensify/App/blob/899b8c7dc6ae8edb4d5d917e3769515d069bfaec/src/libs/actions/PaymentMethods.ts#L301), which means we aren't able to reuse it when we try to reauthenticate. The clearing of the authentication link also seems to be [how we decide to close the the iframe modal in the first place](https://github.com/Expensify/App/blob/899b8c7dc6ae8edb4d5d917e3769515d069bfaec/src/pages/settings/Subscription/CardAuthenticationModal/index.tsx#L59). That's an issue. We need the modal to close without clearing the authentication link from our onyx data. @waterim can you figure out a way to do that?

https://github.com/user-attachments/assets/4160d323-498b-4059-96af-f62f4301862f

**Comment by waterim** [comment]
I will try it tomorrow without clearing the link
Do you think it will be fine to store a secure link all the time?
Or maybe maybe I will remove link after success and keep it after fail, okay, should be fine

**Comment by blimpich** [comment]
If we can clear it on success and keep it on failure then that should be good but I also don't see a problem with keeping it. There's nothing in that link that I'm worried about being exposed on the frontend.

**Comment by blimpich** [comment]
Hmmm, so the iframe opens but then closes before rendering. Not sure why that is.
https://github.com/user-attachments/assets/2c8d2a91-d9d7-4f04-9d8c-03f2f16cf58c

**Comment by blimpich** [comment]
I can make the modal not auto-close by removing [this line](https://github.com/Expensify/App/blob/22d5c0bf1f1d029a8450fed179aedbab7cf92ea7/src/pages/settings/Subscription/CardAuthenticationModal/index.tsx#L43) but I still just get a blank page instead of the stripe verification screen. Probably something backend I need to adjust here. Looking into it.

**Comment by blimpich** [comment]
Ah, I think I've figured it out. Comes down to onyx updates. The old way we did this was we would call `User_VerifySetupIntent` when we wanted to re-authenticate and that would give us a new authentication link, and we'd use that to make this work. We can't just reuse the old link I don't think. We aren't able to get that new link right now because the backend isn't updating the frontend's onyx data to have the authentication link. I will make a PR for that.

Lets do this:
- revert the changes where we make the modal close by clearing the authenticationLink
- instead of having the authenticate payment button navigate to the add payment card screen, lets have it call `User_VerifySetupIntent`. That call should update the authenticationLink onyx data, which should then cause the modal to open up on the frontend.

The changes I'm going to make also mean that we shouldn't need to use makeRequestWithSideEffect [here](https://github.com/Expensify/App/blob/22d5c0bf1f1d029a8450fed179aedbab7cf92ea7/src/libs/actions/PaymentMethods.ts#L258). The backend will update the onyx key and push it to the frontend.

Does that sound good @waterim?

**Comment by waterim** [comment]
Sounds great and makes a lot of sense for me know. thank you!

**Comment by mananjadhav** [comment]
I'll start with the review of this code, but I don't see any screenshots/videos for the testing. @waterim Can you please add them?

**Comment by waterim** [comment]
@mananjadhav Its not possible to test it without a local backend and dev env stripe
To get this you will need to ask @blimpich to share it with you 

Also I will be able to upload a video with steps only when @blimpich ngrok is online (in 3-4 hours from now)

If its okay for you you are able to go through code for now and when @blimpich will be online we can proceed with testing this flow manually

**Comment by waterim** [comment]
@mananjadhav added video with the flow

File: src/pages/settings/Subscription/CardSection/CardSection.tsx [hunk 1]
```diff
@@ -81,7 +95,8 @@ function CardSection() {
         BillingBanner = <TrialStartedBillingBanner />;
     } else if (SubscriptionUtils.hasUserFreeTrialEnded()) {
         BillingBanner = <TrialEndedBillingBanner />;
-    } else if (billingStatus) {
+    }
+    if (billingStatus) {
```
Comment (mananjadhav) [inline src/pages/settings/Subscription/CardSection/CardSection.tsx:99]: Is this `else if` --> `if` intentional?

Are we saying along with Trial/Trial ended banner we would show the authenticate payment button?

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/fb470143-b04d-4122-b632-96b1c6716618

</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by mananjadhav** [comment]
@blimpich There is no popup shown on payment currency change right.


Steps:
1. Go to subscription page.
2. Add a card in USD.
3. Go to the payment method row and lcik on three dots menu.
4. Click change payment currency.
5. Select GBP and enter CVV.
6. It doesn't show any popup like it does when we add a GBP card.

File: src/pages/settings/Subscription/CardSection/CardSection.tsx (hunk 1 above)
Comment (mananjadhav) [reply src/pages/settings/Subscription/CardSection/CardSection.tsx:99]: I think I got the change now.

**Comment by blimpich** [comment]
> There is no popup shown on payment currency change right.

Correct. That is a known issue, choosing not to prioritize right now.

**Review by mananjadhav — APPROVED** [review]
The code changes look good and also works well.

My iOS build is failing but this feature is anyway not available for native app.

Another follow up thing unrelated to the current PR could be, when clicking on Submit instead of Fail, `Authenticate payment` modal doesn't have a good visible loader. This is I think loaded from iframe. The rest it looks good.

**Review by blimpich — CHANGES_REQUESTED** [review]
Looks good and tests well but can we follow up on this comment I made awhile back?

> The changes I'm going to make also mean that we shouldn't need to use makeRequestWithSideEffect [here](https://github.com/Expensify/App/blob/22d5c0bf1f1d029a8450fed179aedbab7cf92ea7/src/libs/actions/PaymentMethods.ts#L258). The backend will update the onyx key and push it to the frontend.

It would be good to get rid of the side effect usage where we can, and I think we can here.

**Comment by blimpich** [comment]
@waterim can you also update this branch with `main` in addition to changes requested above?

**Comment by waterim** [comment]
@blimpich Sure!
I had days off and today Callstack had half-day upskilling day and now Im working on this ticket!

**Comment by waterim** [comment]
@blimpich updated, can you test please?

**Comment by blimpich** [comment]
Ugh, I was wrong, I misunderstood one of the backend api endpoints and how it worked. @waterim can you please revert the changes related to getting rid of `makeRequestWithSideEffects()`? Then lets get this merged. Thank you for working on this and being patient with me!

**Comment by waterim** [comment]
Sure! Will revert it tomorrow morning :)

**Review by blimpich — APPROVED** [review]
Tested it again, works great! Lets merge this!!!

**Comment by mvtglobally** [comment]
Are we able to use this test data in Staging? I think Staging might be pointing to PROD Stripe. 

image (14)
image (13)

https://github.com/user-attachments/assets/5782eee9-d517-4788-a105-72f02eeb7b33

**Pull request merged**